### DISCUSSION TITLE
How to calculate the frequency for SPEI-6 index?

### DISCUSSION THREAD

**Discussion by CGL5230** [Ouranosinc/xclim#1594] [Questions] [answered]
Hi,all! I use SPEI index by xclim.
`SPEI_BCC_CSM2MR_ssp126=xclim.indices.standardized_precipitation_evapotranspiration_index(wb=wb_BCC_CSM2MR_ssp126,freq='MS',window=6)`
 My dataset time period is 2015-2100 and time resolution is daily. When I finished calculating the SPEI, the resolution of the data seems to have changed to be in months. How do I calculate the frequency of SPEI<-1 at this time? Or how do I represent this thing?

I see xclim offer a example to analysis[ frequency](https://xclim.readthedocs.io/en/stable/notebooks/frequency_analysis.html) and the formulate to [drought freqency](https://www.nature.com/articles/s41597-022-01279-5). Is there any way to calculate the frequency for SPEI-6? like：
```
drought_events = (spi6 < drought_threshold).groupby('time').sum() 
total_years = len(precip_data.time.dt.year) 
drought_frequency = (drought_events > 0).sum() / total_years
```
The frequency for drought:
![image](https://github.com/Ouranosinc/xclim/assets/57136380/ba3af488-5cba-4033-b923-107e43dcaedd)

**Comment by aulemahal** [comment, marked as answer]
Hi @CGL5230 !

Your draft has the basic idea. The simplest way I see to make this calculation would be without xclim, with only xarray basics:

```python3
is_drought = (spi6 < drought_threshold)
drought_freq = 100 * is_drought.resample(time='YS').sum() / spi6.notnull().resample(time='YS').sum()
drought_freq.attrs.update(long_name='Drought frequency', units='%')
```
Instead other counting the number of elements in the timeseries directly like you did, I counted the number of "non-null" (valid) values of SPI6. For usual climate simulation data, this makes almost no difference, but it would be more robust in the case of observations, IMO.

Also, I computed the annual drought freq. You could always aggregate to longer periods by averaging.

As you can see, this type of frequency calculation is pretty straightforward. The "frequency analysis" tools of xclim tackle the more complex questions of return periods and exceedance values, fitting your data with known distributions. (Like SPI does internally already).

An example would be:
```python3
xclim.generic.return_level(da=spi6, mode='high', t=20, dist='norm', window=1, freq='YS', method='ML')
```
Which would give you the 20-year return level of annual SPI6. However, my knowledge of SPI is weak and I can't say if that is a pertinent statistic to compute or not.
